frontend/src/metabase/admin/settings/components/ApiKeys/ManageApiKeys.tsx

```diff
@@ -0,0 +1,169 @@
+
+function formatMaskedKey(maskedKey: string) {
+  return maskedKey.substring(0, 4) + "...";
+}
+
+function EmptyTableWarning() {
+  return (
+    <Stack mt="xl" align="center" justify="center" spacing="sm">
+      <Title>{t`No API keys here yet`}</Title>
+      <Text color="text.1">{t`You can create an API key to make API calls programatically.`}</Text>
+    </Stack>
+  );
+}
+
+function ApiKeysTable({
+  apiKeys,
+  setActiveApiKey,
+  setModal,
+  loading,
+  error,
+}: {
+  apiKeys?: ApiKey[];
+  setActiveApiKey: (apiKey: ApiKey) => void;
+  setModal: (modal: Modal) => void;
+  loading: boolean;
+  error?: Error;
+}) {
+  return (
+    <Stack data-testid="api-keys-table" pb="lg">
+      <table className="ContentTable border-bottom">
+        <thead>
+          <tr>
+            <th>{t`Key name`}</th>
+            <th>{t`Group`}</th>
+            <th>{t`Key`}</th>
+            <th>{t`Last Modified By`}</th>
+            <th>{t`Last Modified On`}</th>
+            <th />
+          </tr>
+        </thead>
+        <tbody>
+          {apiKeys?.map(apiKey => (
+            <tr key={apiKey.id} className="border-bottom">
+              <td className="text-bold">{apiKey.name}</td>
+              <td>{apiKey.group.name}</td>
+              <td className="text-monospace">
+                {formatMaskedKey(apiKey.masked_key)}
+              </td>
+              <td>{apiKey.updated_by.common_name}</td>
+              <td>{formatDateTimeWithUnit(apiKey.updated_at, "minute")}</td>
+              <td>
+                <Group spacing="md">
+                  <Icon
+                    name="pencil"
+                    className="cursor-pointer"
+                    onClick={() => {
+                      setActiveApiKey(apiKey);
+                      setModal("edit");
+                    }}
+                  />
+                  <Icon
+                    name="trash"
+                    className="cursor-pointer"
+                    onClick={() => {
+                      setActiveApiKey(apiKey);
+                      setModal("delete");
+                    }}
+                  />
+                </Group>
+              </td>
+            </tr>
+          ))}
+        </tbody>
+      </table>
+      <LoadingAndErrorWrapper loading={loading} error={error}>
+        {apiKeys?.length === 0 && <EmptyTableWarning />}
+      </LoadingAndErrorWrapper>
+    </Stack>
+  );
+}
+
+export const ManageApiKeys = () => {
+  const [modal, setModal] = useState<Modal>(null);
+  const [activeApiKey, setActiveApiKey] = useState<null | ApiKey>(null);
+
+  const [{ value: apiKeys, loading, error }, refreshList] = useAsyncFn(
+    (): Promise<ApiKey[]> => ApiKeysApi.list(),
+    [],
+  );
+
+  const handleClose = () => setModal(null);
+
+  useEffect(() => {
+    refreshList();
+  }, [refreshList]);
+
+  const isShowingEmptyTable = !loading && !error && apiKeys?.length === 0;
+
+  return (
+    <>
+      {modal === "create" ? (
```

Comment:
I think we generally use redux state or routes to handle stuff like this, but but it's not something I'd get too hung up over

Reply:
nonblocking suggestion: To avoid the nested ternaries, some of this logic could be extracted into a `renderModal` function.

Reply:
Not a fan of using global state to handle these little local modals. Extracted them into a `ApiKeysModals` component to encapsulate the modal logic instead of having nested ternaries.